Added websocket client handler, and websocket-client module

antoinetran opened the pull request:
https://github.com/permitio/fastapi_websocket_rpc/issues/31

Tested, seems to work fine in both websockets and websocket!

Will try to add doc later.

antoinetran commented:
Oups, I just saw that exception handling is not clean. Will look into it later.

antoinetran commented:
Ok:

- doc added
- exception handled per Websocket Client Handler (exceptions during read and during connect)
- comments added
- both handlers tested with and without proxy

antoinetran commented:
@orweis I think I am done! Please give me a feedback when you can, and hopefully release a tag if possible :)

orweis commented on fastapi_websocket_rpc/websocket_rpc_client.py:
```diff
@@ -205,12 +333,13 @@ async def reader(self):
-            logger.exception("RPC Reader task failed")
-            raise
+        except Exception as err:
+            if await self.ws.isConnectionClosedException(err):
```
Would be better to use exception cases instead of if clauses (more efficient) , this can be achieved by taking the entire function into polymorphism

antoinetran replied:
Done! Created BaseConnectionClosedException

orweis replied:
Sorry for dealyed reposnses, this week is very busy. I've cleared time to review next week.

antoinetran replied:
Hi @orweis , maybe if you have just an example, then I could adapt the code? Thanks!

orweis replied:
Hi, sorry for dropping the ball here, this last week has been extremely busy with end of year around the corner.

I'll do my best to provide you with more guidence later today

antoinetran replied:
I do not see how to apply your guidance below (putting exception handling in connect()) here. If we put exception handling inside recv(), then we wouldn't be able to get out of the while loop. Or maybe you want each websocket client handler to implement the entire reader() method? Then there would be a bit of duplicate code. I will try that today.

orweis replied:
Good point re- recv() / reader() exceptions .
How about having a shared _recv() and have each Class in its recv() call it and catch its exceptions and raise instead a common shared exceptions  (e.g. a joint ConnectionClosed)

antoinetran replied:
I have been thinking a bit about it... To my mind, I find it ok to have a reraise, in case of connection closed, because it performance-wise, it only happens once, then the framework close channel, etc.
But anyway, I think about simplicity and modified the code, so that we only expect a "None" as a special return result, in case of connection closed. No need to reraise, no need for a special base exception!

antoinetran replied:
https://github.com/antoinetran/fastapi_websocket_rpc/commit/7d899469f23299b7e808729acd0bbfba7a9d0143

orweis reviewed (changes requested):
Hi @antoinetran  - this is greta - kudos!
I did an initial review - and there's some good stuff to start here

antoinetran commented on fastapi_websocket_rpc/websocket_rpc_client.py:
```diff
@@ -205,12 +333,13 @@ async def reader(self):
-            logger.exception("RPC Reader task failed")
-            raise
+        except Exception as err:
+            if self.ws.isConnectionClosedException(err):
```
But we don't expect the exception loop to be reach lots of time? I am not sure what you want. Is-it:
```
        except Exception as err:
            self.ws.handleExceptionDuringRead(err):
```

antoinetran replied:
Ok this remark is duplicated below (outdated). I am closing this one in favor of below.

antoinetran commented on requirements.txt:
```diff
@@ -3,4 +3,5 @@ packaging>=20.4
 pydantic>=1.9.1
 uvicorn>=0.17.6,<1
 websockets>=10.3,<11
+websocket-client>=1.6.4
```
Done. Added info in README

orweis commented on fastapi_websocket_rpc/websocket_rpc_client.py:
```diff
@@ -15,6 +15,149 @@
+
+    async def handle_exception(self, exception: Exception):
+        try:
+            raise exception
```
This is actually worse performance-wise - catching and then re raising the exception.
What I meant was to have the __connect__ method be polymorphic (or just a portion of it)

orweis replied:
So instead of having the `handle_exception` method which starts with `raise exception`.

You have the polymorphism in the `connect` method itself -


```python

class WebSocketsClientHandler(SimpleWebSocket):
    #.... rest of code
    
    # Connect with type specific exceptions
    async def connect(self, uri: str, **connect_kwargs):
        try:
            self._websocket = await websockets.connect(uri, **connect_kwargs)
        except ConnectionRefusedError:
            logger.info("RPC connection was refused by server")
            raise
        except ConnectionClosedError:
            logger.info("RPC connection lost")
            raise
        except ConnectionClosedOK:
            logger.info("RPC connection closed")
            raise
        except InvalidStatusCode as err:
            logger.info(
                f"RPC Websocket failed - with invalid status code {err.status_code}")
            raise
        except WebSocketException as err:
            logger.info(f"RPC Websocket failed - with {err}")
            raise  

class WebSocketClientHandler(SimpleWebSocket):
    #.... rest of code
    
    # Connect with type specific exceptions
    async def connect(self, uri: str, **connect_kwargs):
        try:
           self._websocket = await asyncio.get_event_loop().run_in_executor(None, websocket.create_connection, uri, **connect_kwargs)
        # See https://websocket-client.readthedocs.io/en/latest/exceptions.html
        except websocket._exceptions.WebSocketAddressException:
            logger.info("websocket address info cannot be found")
            raise
        except websocket._exceptions.WebSocketBadStatusException:
            logger.info("bad handshake status code")
            raise
        except websocket._exceptions.WebSocketConnectionClosedException:
            logger.info("remote host closed the connection or some network error happened")
            raise
        except websocket._exceptions.WebSocketPayloadException:
            logger.info(
                f"WebSocket payload is invalid")
            raise
        except websocket._exceptions.WebSocketProtocolException:
            logger.info(f"WebSocket protocol is invalid")
            raise
        except websocket._exceptions.WebSocketProxyException:
            logger.info(f"proxy error occurred")
            raise


# With the generic excpetions (OSError , Exception) being handled in the wrapping  Class - WebSocketRpcClient

async def __connect__(self):
        try:
           # .... actual code

        except OSError as err:
            logger.info("RPC Connection failed - %s", err)
            raise
        except Exception as err:
            logger.exception("RPC Error")
            raise

```

antoinetran replied:
Oh I see, you meant to put the exception handling inside the connect function. Yes no problem with that! Will do.

antoinetran replied:
Done!

antoinetran replied:
https://github.com/antoinetran/fastapi_websocket_rpc/commit/54e12efd4353ffa1a0f94bbc97e041385f9771c4

orweis commented on fastapi_websocket_rpc/websocket_rpc_client.py:
```diff
@@ -15,6 +15,149 @@
+            logger.exception("RPC Error")
+            raise
+
+class WebSocketsClientHandler(SimpleWebSocket):
```
Maybe we should give these a better name  - the difference of just one letter "S" can be confusing.
Maybe `ProxyEnabledWebSocketClientHandler` ?

antoinetran replied:
I thought there might be other reasons for users to switch to websocket-client, other than proxy. I don't know websocket-client enough. But ok to rename it. Done!

antoinetran replied:
https://github.com/antoinetran/fastapi_websocket_rpc/commit/385df4392aca04a49ff934ed349f8f3eb1b30495

orweis replied:
Cool - @antoinetran :) 

@roekatz can you give this an additional look - this being sensitive code - it could use another eye

roekatz commented on fastapi_websocket_rpc/websocket_rpc_client.py:
```diff
@@ -137,25 +280,6 @@ async def __connect__(self):
-            logger.info("RPC Connection failed - %s", err)
-            raise
         except Exception as err:
             logger.exception("RPC Error")
```
When `connect` raises an error - the exception (with stack trace) is logged twice

orweis replied:
Nice catch.
@antoinetran - one last fix and I think we can merge! :) 

P.s.
Can you DM me / email me (or@permit.io) your shipping address - will send you some swag- if you want :)

antoinetran replied:
This is fixed: https://github.com/antoinetran/fastapi_websocket_rpc/commit/3e985f6f0e43746daff64860ee3f2473f3fd382a

roekatz reviewed (changes requested):
Other than the double logging of exceptions in `connect` - Looks great to me :)

antoinetran commented:
Hello! Please wait for me to do final tests. I found some minor bugs.

antoinetran commented:
Ok now I tested in my small app, both in server and client code (before I only tested on client). I fixed an error on server-side (endpoint) in WebSocketSimplifier, because it implements SimpleWebSocket too, It works fine! Normally this is it.

Can you accept the test workflow here? https://github.com/permitio/fastapi_websocket_rpc/actions/runs/7400381788 There was an error because it was missing an os import (fixed).

Then, I don't know what is the merge process. Do you need me to squash the history, or do you merge as-it?

antoinetran commented:
@orweis  I have the impression https://github.com/permitio/fastapi_websocket_rpc/actions/runs/7400381788/job/20134732113#step:6:80 failed because of incorrect `rootdir`:
```
rootdir: /home/runner/work/fastapi_websocket_rpc/fastapi_websocket_rpc
```
should be
```
rootdir: /home/runner/work/fastapi_websocket_rpc
```
Then, the `ModuleNotFoundError: No module named 'fastapi_websocket_rpc.rpc_methods'`error would disappear, because it was already in fastapi_websocket_rpc directory. But I don't know what is causing the wrong rootdir. Looking into it...

orweis commented:
Maybe try to fix the import, try relative import with . Synatx , absolute with adding the parent folder to sys path

antoinetran commented:
> Maybe try to fix the import, try relative import with . Synatx , absolute with adding the parent folder to sys path

Ok I read lots of documentation. Can you try running https://github.com/permitio/fastapi_websocket_rpc/actions/runs/7401903856? It seems pytest version has updated, which caused a regression. I don't have experience with pytest, but I tried a few solution and this commit seems to fix it: https://github.com/antoinetran/fastapi_websocket_rpc/commit/91d8d80c68fb972859e1e478d297b24a7de9fcb9

antoinetran commented:
Ok there were some issues regarding Unit Tests:
* somehow, the module cannot be found anymore, I think this is because of pytest update? I fixed it by writing to pytest.ini the `pythonpath = .`, and removed some PATH hack in the test files. This seems to work now.
* the unit test were running indefinitely (6 hours), nothing is shown. It took me a while to figure out, by adding verbose + capture stdout that one Unit Test implements the SimpleWebSocket too and thus was missing the new method.

Still working on it.

orweis commented:
No worries, take your time. And thank you

orweis commented:
This looks good, @roekatz your final review here is requested.
